ext/drreg/drreg.c

```diff
@@ -531,6 +536,11 @@ drreg_insert_restore_all(void *drcontext, instrlist_t *bb, instr_t *inst,
                 (instr_is_label(inst) &&
                  ((ptr_uint_t)instr_get_note(inst) == DR_NOTE_ANNOTATION ||
                   (ptr_uint_t)instr_get_note(inst) == DR_NOTE_REG_BARRIER)) ||
+                /* i#6417: in case of a syscall, restore the values of registers since
+                 * they may be used as input parameters for the kernel. For example,
+                 * ECX is used to store the length for mmap2.
+                 */
+                instr_is_syscall(inst) ||
```

Comment:
This looks good for pre-syscall.  But after the syscall we need to update the spilled values with whatever the kernel wrote: so we need to also check for instr_is_syscall in the loop labeled `/* After each app write, update spilled app values: */`.  Otherwise, a restore at the end of the block will bring in the old pre-syscall value.

To test, add an asm sequence to drreg-test (ifdef LINUX I guess) with a single block with a syscall in the middle and a pattern the test client can find (see existing DRREG_TEST_38_ASM e.g.) and the test client for that block can then reserve eax and ecx and clobber both register both before and after the syscall. 
 Verify the mmap succeeded and that the app's eax is correctly restored at the end of the block (maybe have the app check the syscall return value in the next block or sthg).